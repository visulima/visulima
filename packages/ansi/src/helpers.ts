// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
const isBrowser = globalThis?.window?.document !== undefined;

export const isTerminalApp = !isBrowser && process.env.TERM_PROGRAM === "Apple_Terminal";
export const isWindows = !isBrowser && process.platform === "win32";
