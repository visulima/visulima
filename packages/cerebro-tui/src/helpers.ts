import cliTruncate from "cli-truncate";
import stringWidth from "string-width";
import wordwrap from "wordwrap";

/**
 * Applies padding to the left or the right of the string by repeating
 * a given char.
 *
 * The method is not same as `padLeft` or `padRight` from JavaScript STD lib,
 * since it repeats a char regardless of the max width.
 */
const applyPadding = (value: string, options: { paddingChar: string; paddingLeft?: number; paddingRight?: number }): string => {
    if (options.paddingLeft) {
        // eslint-disable-next-line no-param-reassign
        value = `${options.paddingChar.repeat(options.paddingLeft)}${value}`;
    }

    if (options.paddingRight) {
        // eslint-disable-next-line no-param-reassign
        value = `${value}${options.paddingChar.repeat(options.paddingRight)}`;
    }

    return value;
};

/**
 * Justify the columns to have the same width by filling
 * the empty slots with a padding char.
 *
 * Optionally, the column can be aligned left or right.
 */
export const justify = (
    columns: string[],
    options: {
        align?: "left" | "right";
        maxWidth: number;
        paddingChar?: string;
    },
): string[] => {
    const normalizedOptions = {
        align: "left" as const,
        paddingChar: " ",
        ...options,
    };

    return columns.map((column) => {
        const columnWidth = stringWidth(column);

        /**
         * Column is already same or greater than the maxWidth
         */
        if (columnWidth >= normalizedOptions.maxWidth) {
            return column;
        }

        /**
         * Fill empty space on the right
         */
        if (normalizedOptions.align === "left") {
            return applyPadding(column, {
                paddingChar: normalizedOptions.paddingChar,
                paddingRight: normalizedOptions.maxWidth - columnWidth,
            });
        }

        /**
         * Fill empty space on the left
         */
        return applyPadding(column, {
            paddingChar: normalizedOptions.paddingChar,
            paddingLeft: normalizedOptions.maxWidth - columnWidth,
        });
    });
};

/**
 * Wrap the text under the starting and the ending column.
 * The first line will start at 1st column. However, from
 * the 2nd line onwards, the columns before the start
 * column are filled with white space.
 */
export const wrap = (
    columns: string[],
    options: {
        endColumn: number;
        startColumn: number;
        trimStart?: boolean;
    },
): string[] => {
    const wrapper = wordwrap(options.startColumn, options.endColumn);

    if (options.trimStart) {
        return columns.map((column) => wrapper(column).trimStart());
    }

    return columns.map((column) => wrapper(column));
};

/**
 * Truncates the text after a certain width.
 */
export const truncate = (
    columns: string[],
    options: {
        maxWidth: number;
        position?: "end" | "middle" | "start";
        truncationChar?: string;
    },
): string[] =>
    columns.map((column) =>
        cliTruncate(column, options.maxWidth, {
            position: options.position ?? "end",
            truncationCharacter: options.truncationChar ?? "â€¦",
        }),
    );
