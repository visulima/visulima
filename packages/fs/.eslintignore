dist
node_modules
coverage
vitest.config.ts
recipes
__fixtures__
__bench__
.prettierrc.cjs
.eslintrc.cjs
.secretlintrc.cjs
_meta.*.json

packem.config.ts
package.json
README.md
