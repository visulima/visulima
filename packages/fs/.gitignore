__fixtures__/find-up/file-link
