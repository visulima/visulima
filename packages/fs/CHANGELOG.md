## @visulima/fs [2.1.5](https://github.com/visulima/visulima/compare/@visulima/fs@2.1.4...@visulima/fs@2.1.5) (2024-06-11)

### Bug Fixes

* **fs:** updated type-fest to v4.20.0 ([b34f417](https://github.com/visulima/visulima/commit/b34f417537b209f4c411103b6eef0e5a6d9b947d))

### Build System

* fixed found audit error, updated all dev package deps, updated deps in apps and examples ([4c51950](https://github.com/visulima/visulima/commit/4c519500dc5504579d35725572920658999885cb))

## @visulima/fs [2.1.4](https://github.com/visulima/visulima/compare/@visulima/fs@2.1.3...@visulima/fs@2.1.4) (2024-06-06)


### Bug Fixes

* allow node v22 ([890d457](https://github.com/visulima/visulima/commit/890d4570f18428e2463944813c0c638b3f142803))



### Dependencies

* **@visulima/path:** upgraded to 1.0.2
* **@visulima/error:** upgraded to 3.2.11

## @visulima/fs [2.1.3](https://github.com/visulima/visulima/compare/@visulima/fs@2.1.2...@visulima/fs@2.1.3) (2024-06-05)


### Bug Fixes

* **fs:** updated type-fest ([f1ba811](https://github.com/visulima/visulima/commit/f1ba81100ad6eddeeda80066b2fb0f0b57a4cfa5))


### Miscellaneous Chores

* **fs:** added test for dot file ([105c6f3](https://github.com/visulima/visulima/commit/105c6f329312c71d3596b7115c16eb3829d7eac9))



### Dependencies

* **@visulima/error:** upgraded to 3.2.10

## @visulima/fs [2.1.2](https://github.com/visulima/visulima/compare/@visulima/fs@2.1.1...@visulima/fs@2.1.2) (2024-05-24)


### Bug Fixes

* changed pathe to @visulima/path ([#410](https://github.com/visulima/visulima/issues/410)) ([bfe1287](https://github.com/visulima/visulima/commit/bfe1287aff6d28d5dca302fd4d58c1f6234ce0bb))


### Miscellaneous Chores

* changed semantic-release-npm to pnpm ([b6d100a](https://github.com/visulima/visulima/commit/b6d100a2bf3fd026577be48726a37754947f0973))
* fixed wrong named folders to integration, added TEST_PROD_BUILD ([1b826f5](https://github.com/visulima/visulima/commit/1b826f5baf8285847199de9ede8fbdbadf201ad6))



### Dependencies

* **@visulima/path:** upgraded to 1.0.1
* **@visulima/error:** upgraded to 3.2.9

## @visulima/fs [2.1.1](https://github.com/visulima/visulima/compare/@visulima/fs@2.1.0...@visulima/fs@2.1.1) (2024-05-15)



### Dependencies

* **@visulima/error:** upgraded to 3.2.8

## @visulima/fs [2.1.0](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.9...@visulima/fs@2.1.0) (2024-05-06)


### Features

* **fs:** adding ensure symlink ([#350](https://github.com/visulima/visulima/issues/350)) ([3e2fd19](https://github.com/visulima/visulima/commit/3e2fd191c4de13a2a6ebf599cd83926b7f8ad9a6))

## @visulima/fs [2.0.9](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.8...@visulima/fs@2.0.9) (2024-05-03)


### Bug Fixes

* updated type-fest to version ^4.18.1 ([ab96053](https://github.com/visulima/visulima/commit/ab9605355b6d84d1662fa19ef919682c33e68b5b))

## @visulima/fs [2.0.8](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.7...@visulima/fs@2.0.8) (2024-04-27)


### Bug Fixes

* **deps:** updated type-fest dep ([93445aa](https://github.com/visulima/visulima/commit/93445aaa7c8a5431b3ecffe5bfb90a54f635d68d))



### Dependencies

* **@visulima/error:** upgraded to 3.2.7

## @visulima/fs [2.0.7](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.6...@visulima/fs@2.0.7) (2024-04-17)



### Dependencies

* **@visulima/error:** upgraded to 3.2.6

## @visulima/fs [2.0.6](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.5...@visulima/fs@2.0.6) (2024-04-09)



### Dependencies

* **@visulima/error:** upgraded to 3.2.5

## @visulima/fs [2.0.5](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.4...@visulima/fs@2.0.5) (2024-04-05)


### Bug Fixes

* updated type-fest and dev deps ([d7f648d](https://github.com/visulima/visulima/commit/d7f648debdb10eeeb4b8942c45b2e6f6ead560e2))

## @visulima/fs [2.0.4](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.3...@visulima/fs@2.0.4) (2024-04-02)


### Bug Fixes

* **fs:** fixed default behavior of collect ([#387](https://github.com/visulima/visulima/issues/387)) ([172baa9](https://github.com/visulima/visulima/commit/172baa9ebd724c1d145e78f926ec12d068ef73bb))

## @visulima/fs [2.0.3](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.2...@visulima/fs@2.0.3) (2024-03-30)


### Bug Fixes

* updated dev dependencies and type-fest ([1df1a35](https://github.com/visulima/visulima/commit/1df1a353f0bc7968087d5c68e0712c262837535b))



### Dependencies

* **@visulima/error:** upgraded to 3.2.4

## @visulima/fs [2.0.2](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.1...@visulima/fs@2.0.2) (2024-03-27)


### Bug Fixes

* added missing os key to package.json ([4ad1268](https://github.com/visulima/visulima/commit/4ad1268ed12cbdcf60aeb46d4c052ed1696bc150))



### Dependencies

* **@visulima/error:** upgraded to 3.2.3

## @visulima/fs [2.0.1](https://github.com/visulima/visulima/compare/@visulima/fs@2.0.0...@visulima/fs@2.0.1) (2024-03-24)


### Bug Fixes

* changed parse back to node:path ([#371](https://github.com/visulima/visulima/issues/371)) ([675dd0a](https://github.com/visulima/visulima/commit/675dd0a39b6cd3c4559472608b23c55196df12a0))

## @visulima/fs [2.0.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.11.1...@visulima/fs@2.0.0) (2024-03-23)


### âš  BREAKING CHANGES

* **fs:** moved node:path to pathe and changed NotFoundError message

Signed-off-by: prisis <d.bannert@anolilab.de>

### Features

* **fs:** moved node:path to pathe, changed NotFoundError message ([0557a18](https://github.com/visulima/visulima/commit/0557a18d97af6061ba9cb1581c420c2bc72965ef))

## @visulima/fs [1.11.1](https://github.com/visulima/visulima/compare/@visulima/fs@1.11.0...@visulima/fs@1.11.1) (2024-03-22)


### Bug Fixes

* **fs:** updated type-fest ([aa84569](https://github.com/visulima/visulima/commit/aa84569aca39f2b3af3508fbae9a56ea2da33b9e))



### Dependencies

* **@visulima/error:** upgraded to 3.2.2

## @visulima/fs [1.11.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.10.0...@visulima/fs@1.11.0) (2024-03-20)


### Features

* **fs:** adding ensure methods ([#355](https://github.com/visulima/visulima/issues/355)) ([0456bc7](https://github.com/visulima/visulima/commit/0456bc7753d8a117e5bab23d74000906ba1d98ba))

## @visulima/fs [1.10.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.9.0...@visulima/fs@1.10.0) (2024-03-19)


### Features

* **fs:** added remove a wrapper around rm and unlink ([#354](https://github.com/visulima/visulima/issues/354)) ([ea30c3e](https://github.com/visulima/visulima/commit/ea30c3e3efd75e2bf07ca81af701543e087aff25))

## @visulima/fs [1.9.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.8.0...@visulima/fs@1.9.0) (2024-03-19)


### Features

* **fs:** added read and write yaml functions ([#351](https://github.com/visulima/visulima/issues/351)) ([365b1ed](https://github.com/visulima/visulima/commit/365b1ed5cbf33c356e4427fb4fd8f2d93ee14733))

## @visulima/fs [1.8.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.7.1...@visulima/fs@1.8.0) (2024-03-17)


### Features

* **fs:** added eol, emptyDir and emptyDirSync ([#348](https://github.com/visulima/visulima/issues/348)) ([8084200](https://github.com/visulima/visulima/commit/8084200779f671403b9585c3b6fec985641c1cf6))

## @visulima/fs [1.7.1](https://github.com/visulima/visulima/compare/@visulima/fs@1.7.0...@visulima/fs@1.7.1) (2024-03-16)



### Dependencies

* **@visulima/error:** upgraded to 3.2.1

## @visulima/fs [1.7.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.6.0...@visulima/fs@1.7.0) (2024-03-16)


### Features

* **package:** added cjs and esm export ([#335](https://github.com/visulima/visulima/issues/335)) ([62b873a](https://github.com/visulima/visulima/commit/62b873a76d2a5356e7596d3a7e299d8e51eb2fcd))

## @visulima/fs [1.6.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.5.0...@visulima/fs@1.6.0) (2024-03-14)


### Features

* **fs:** added matcher and FIND_UP_STOP to findup ([#344](https://github.com/visulima/visulima/issues/344)) ([34a5844](https://github.com/visulima/visulima/commit/34a584450416b8e99b70b920357cbde57e2c6b5b))


### Bug Fixes

* **fs:** added missing constants exports ([51915b2](https://github.com/visulima/visulima/commit/51915b257fa416c7fbcca24f16eb98132273cf4c))

## @visulima/fs [1.5.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.4.0...@visulima/fs@1.5.0) (2024-03-14)


### Features

* **fs:** added findUp array support, based on the other the file will be found ([#343](https://github.com/visulima/visulima/issues/343)) ([4816318](https://github.com/visulima/visulima/commit/4816318770049c4d9a9cfa27c251449ce3077e8f))
* **fs:** added new toPath util ([d65cf3e](https://github.com/visulima/visulima/commit/d65cf3e2f53a4d81f927740050741d59bcccfc9f))


### Bug Fixes

* **fs:** fixed wrong return type on findUp ([f3c96de](https://github.com/visulima/visulima/commit/f3c96def18cd8e70093bdd943c2e2f396436dda4))

## @visulima/fs [1.4.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.3.1...@visulima/fs@1.4.0) (2024-03-11)


### Features

* **fs:** added new stringify option to write json ([5549b80](https://github.com/visulima/visulima/commit/5549b80386aa8ea5f47997dea50dc05209245008))

## @visulima/fs [1.3.1](https://github.com/visulima/visulima/compare/@visulima/fs@1.3.0...@visulima/fs@1.3.1) (2024-03-11)


### Bug Fixes

* **fs:** exported the find-up options as type ([f8b2321](https://github.com/visulima/visulima/commit/f8b2321cead794003f5100d2aa8b30b31d42cc64))

## @visulima/fs [1.3.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.2.1...@visulima/fs@1.3.0) (2024-03-11)


### Features

* **fs:** add read and write for file and json ([#341](https://github.com/visulima/visulima/issues/341)) ([a2e16d8](https://github.com/visulima/visulima/commit/a2e16d88ea6b14cef1b55ff7ee5d6a9e4afe262f))

## @visulima/fs [1.2.1](https://github.com/visulima/visulima/compare/@visulima/fs@1.2.0...@visulima/fs@1.2.1) (2024-03-09)


### Bug Fixes

* added missing type module to the package.json ([510c5b7](https://github.com/visulima/visulima/commit/510c5b7e9cdca2b6de104d0b6b0f5ad2fbf50956))

## @visulima/fs [1.2.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.1.0...@visulima/fs@1.2.0) (2024-03-09)


### Features

* added readFile and readFileSync ([#340](https://github.com/visulima/visulima/issues/340)) ([47dd800](https://github.com/visulima/visulima/commit/47dd800cbadc8d353e9546487fa4c288d3f8be79))

## @visulima/fs [1.1.0](https://github.com/visulima/visulima/compare/@visulima/fs@1.0.0...@visulima/fs@1.1.0) (2024-03-09)


### Features

* **fs:** added isAccessibleSync and isAccessible ([#339](https://github.com/visulima/visulima/issues/339)) ([552eaec](https://github.com/visulima/visulima/commit/552eaec4a1b36db08a6386c562fa6a3bf248c40d))

## @visulima/fs 1.0.0 (2024-03-09)


### Features

* **fs:** added collect-sync ([ed76052](https://github.com/visulima/visulima/commit/ed760529d190a177c64c4028c572160da62c9a0f))
* migrating readdir to fs package ([664d567](https://github.com/visulima/visulima/commit/664d5670ac84811963a830b594968f19bf81809a))
* renamed readdir to fs ([29b0564](https://github.com/visulima/visulima/commit/29b0564913cd136cd126a1e3a6d9461a85158930))


### Bug Fixes

* **fs:** fixed test and lint errors ([01e5f92](https://github.com/visulima/visulima/commit/01e5f921fd2a8700f40464bdc77b0a9ea00ab613))
