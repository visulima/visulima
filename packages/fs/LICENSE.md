MIT License

Copyright (c) 2024 visulima

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

<!-- DEPENDENCIES -->

# Licenses of bundled dependencies

The published @visulima/fs artifact additionally contains code with the following licenses:
MIT

# Bundled dependencies:

## @visulima/error

License: MIT
By: Daniel Bannert
Repository: https://github.com/visulima/visulima.git

> MIT License
>
> Copyright (c) 2024 visulima
>
> Permission is hereby granted, free of charge, to any person obtaining a copy
> of this software and associated documentation files (the "Software"), to deal
> in the Software without restriction, including without limitation the rights
> to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
> copies of the Software, and to permit persons to whom the Software is
> furnished to do so, subject to the following conditions:
>
> The above copyright notice and this permission notice shall be included in all
> copies or substantial portions of the Software.
>
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
> IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
> FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
> AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
> LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
> OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
> SOFTWARE.

---

## detect-indent

License: MIT
By: Sindre Sorhus
Repository: sindresorhus/detect-indent

> MIT License
>
> Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (https://sindresorhus.com)
>
> Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
>
> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
>
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

<!-- DEPENDENCIES -->

<!-- TYPE_DEPENDENCIES -->

# Licenses of bundled types

The published @visulima/fs artifact additionally contains code with the following licenses:
(MIT OR CC0-1.0)

# Bundled types:

## type-fest

License: (MIT OR CC0-1.0)
By: Sindre Sorhus
Repository: sindresorhus/type-fest

<!-- TYPE_DEPENDENCIES -->
