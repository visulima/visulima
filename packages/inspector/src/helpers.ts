import type { Inspect, Options } from "./types";

const ansiColors = {
    // 21-29 are resets for 1-9
    black: ["30", "39"],
    blue: ["34", "39"],
    bold: ["1", "22"],
    brightblack: ["30;1", "39"],
    brightblue: ["34;1", "39"],
    brightcyan: ["36;1", "39"],
    brightgreen: ["32;1", "39"],
    // 10-20 are fonts
    brightmagenta: ["35;1", "39"],
    brightred: ["31;1", "39"],
    brightwhite: ["37;1", "39"],
    brightyellow: ["33;1", "39"],
    cyan: ["36", "39"],
    dim: ["2", "22"],
    green: ["32", "39"],
    grey: ["90", "39"],

    hidden: ["8", "28"],
    // 5 & 6 are blinking
    inverse: ["7", "27"],
    italic: ["3", "23"],
    magenta: ["35", "39"],
    red: ["31", "39"],
    strike: ["9", "29"],
    underline: ["4", "24"],
    white: ["37", "39"],

    yellow: ["33", "39"],
} as const;

const styles: Record<string, keyof typeof ansiColors> = {
    bigint: "yellow",
    boolean: "yellow",
    date: "magenta",
    null: "bold",
    number: "yellow",
    regexp: "red",
    special: "cyan",
    string: "green",
    symbol: "green",
    undefined: "grey",
} as const;

export const truncator = "â€¦";

type AnsiValue = (typeof ansiColors)[keyof typeof ansiColors];

function colorise<S extends string>(value: S, styleType: string): string {
    const color: AnsiValue | undefined = ansiColors[styles[styleType]] || ansiColors[styleType as keyof typeof ansiColors] || "";

    if (!color) {
        return String(value) as S;
    }

    return `\u001B[${color[0]}m${String(value) as S}\u001B[${color[1]}m`;
}

export function normaliseOptions(
    {
        breakLength = Number.POSITIVE_INFINITY,
        colors = false,
        customInspect = true,
        depth = 2,
        maxArrayLength = Number.POSITIVE_INFINITY,
        seen = [],
        showHidden = false,
        showProxy = false,
        stylize = String,

        truncate = Number.POSITIVE_INFINITY,
    }: Partial<Options> = {},
    inspect: Inspect,
): Options {
    const options = {
        breakLength: Number(breakLength),
        colors: Boolean(colors),
        customInspect: Boolean(customInspect),
        depth: Number(depth),
        inspect,
        maxArrayLength: Number(maxArrayLength),
        seen,
        showHidden: Boolean(showHidden),
        showProxy: Boolean(showProxy),
        stylize,
        truncate: Number(truncate),
    };
    if (options.colors) {
        options.stylize = colorise;
    }
    return options;
}

function isHighSurrogate(char: string): boolean {
    return char >= "\uD800" && char <= "\uDBFF";
}

export function truncate(string: number | string, length: number, tail: typeof truncator = truncator) {
    string = String(string);
    const tailLength = tail.length;
    const stringLength = string.length;
    if (tailLength > length && stringLength > tailLength) {
        return tail;
    }
    if (stringLength > length && stringLength > tailLength) {
        let end = length - tailLength;
        if (end > 0 && isHighSurrogate(string[end - 1])) {
            end -= 1;
        }
        return `${string.slice(0, end)}${tail}`;
    }
    return string;
}

export function inspectList(list: ArrayLike<unknown>, options: Options, inspectItem?: Inspect, separator = ", "): string {
    inspectItem = inspectItem || options.inspect;
    const size = list.length;
    if (size === 0) return "";
    const originalLength = options.truncate;
    let output = "";
    let peek = "";
    let truncated = "";
    for (let index = 0; index < size; index += 1) {
        const last = index + 1 === list.length;
        const secondToLast = index + 2 === list.length;
        truncated = `${truncator}(${list.length - index})`;
        const value = list[index];

        // If there is more than one remaining we need to account for a separator of `, `
        options.truncate = originalLength - output.length - (last ? 0 : separator.length);
        const string = peek || inspectItem(value, options) + (last ? "" : separator);
        const nextLength = output.length + string.length;
        const truncatedLength = nextLength + truncated.length;

        // If this is the last element, and adding it would
        // take us over length, but adding the truncator wouldn't - then break now
        if (last && nextLength > originalLength && output.length + truncated.length <= originalLength) {
            break;
        }

        // If this isn't the last or second to last element to scan,
        // but the string is already over length then break here
        if (!last && !secondToLast && truncatedLength > originalLength) {
            break;
        }

        // Peek at the next string to determine if we should
        // break early before adding this item to the output
        peek = last ? "" : inspectItem(list[index + 1], options) + (secondToLast ? "" : separator);

        // If we have one element left, but this element and
        // the next takes over length, the break early
        if (!last && secondToLast && truncatedLength > originalLength && nextLength + peek.length > originalLength) {
            break;
        }

        output += string;

        // If the next element takes us to length -
        // but there are more after that, then we should truncate now
        if (!last && !secondToLast && nextLength + peek.length >= originalLength) {
            truncated = `${truncator}(${list.length - index - 1})`;
            break;
        }

        truncated = "";
    }
    return `${output}${truncated}`;
}

function quoteComplexKey(key: string): string {
    if (/^[a-z_]\w*$/i.test(key)) {
        return key;
    }
    return JSON.stringify(key)
        .replaceAll("'", "\\'")
        .replaceAll('\\"', '"')
        .replaceAll(/(^"|"$)/g, "'");
}

export function inspectProperty([key, value]: [unknown, unknown], options: Options): string {
    options.truncate -= 2;
    if (typeof key === "string") {
        key = quoteComplexKey(key);
    } else if (typeof key !== "number") {
        key = `[${options.inspect(key, options)}]`;
    }
    options.truncate -= (key as string).length;
    value = options.inspect(value, options);
    return `${key}: ${value}`;
}
