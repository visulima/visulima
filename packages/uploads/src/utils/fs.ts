import { promises as fsp } from "node:fs";
import { dirname } from "node:path";

/**
 * Ensures that the file exists and returns it size
 * @param path - filename or path to a local file
 * @param overwrite - force creating new empty file
 * @returns file size
 */
export async function ensureFile(path: string, overwrite = false): Promise<number> {
    await fsp.mkdir(dirname(path), { recursive: true });

    const open = await fsp.open(path, overwrite ? "w" : "a");
    await open.close();

    const stat = await fsp.stat(path);

    return stat.size;
}

/**
 * Removes the specified file from the local file system
 */
export async function removeFile(path: string): Promise<void> {
    if (fsp.rm) {
        return fsp.rm(path, { force: true });
    }

    return fsp.unlink(path).catch((error: NodeJS.ErrnoException): void => {
        if (error.code !== "ENOENT") throw error;
    });
}

/**
 * @internal
 */
export { promises as fsp } from "node:fs";
