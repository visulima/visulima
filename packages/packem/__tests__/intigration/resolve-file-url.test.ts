import { rm } from "node:fs/promises";

import { readFileSync, writeFileSync, writeJsonSync } from "@visulima/fs";
import { temporaryDirectory } from "tempy";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { execPackemSync, getNodePathList, streamToString } from "../helpers";

describe.each(await getNodePathList())("node %s - resolve-file-url", (_, nodePath) => {
    let distribution: string;

    beforeEach(async () => {
        distribution = temporaryDirectory();
    });

    afterEach(async () => {
        await rm(distribution, { recursive: true });
    });

    it("should resolve import with file:// annotation", async () => {
        expect.assertions(3);

        writeFileSync(
            `${distribution}/src/importee.mjs`,
            `function log() {
  return 'this should be in final bundle'
}

export default log`,
        );
        writeFileSync(`${distribution}/src/importer.mjs`, `export { default as effect } from "file://${distribution}/src/importee.mjs"`);
        writeJsonSync(`${distribution}/package.json`, {
            main: "./dist/importer.cjs",
            type: "commonjs",
        });

        const binProcess = execPackemSync(["--env NODE_ENV=development"], {
            cwd: distribution,
            nodePath,
        });

        await expect(streamToString(binProcess.stderr)).resolves.toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(`${distribution}/dist/importer.mjs`);

        expect(mjsContent).toBe(`function log() {
  return 'this should be in final bundle'
}

export { log as effect };
`)

        const cjsContent = readFileSync(`${distribution}/dist/importer.cjs`);

        expect(cjsContent).toBe(`'use strict';

function log() {
  return 'this should be in final bundle'
}

exports.effect = log;
`)
    });
});
