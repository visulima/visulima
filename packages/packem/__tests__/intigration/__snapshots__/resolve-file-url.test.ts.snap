// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`node 18.20.1 - resolve-file-url > should resolve import with file:// annotation 1`] = `
"function log() {
  return 'this should be in final bundle'
}

export { log as effect };
"
`;

exports[`node 18.20.1 - resolve-file-url > should resolve import with file:// annotation 2`] = `
"'use strict';

function log() {
  return 'this should be in final bundle'
}

exports.effect = log;
"
`;

exports[`node 20.12.1 - resolve-file-url > should resolve import with file:// annotation 1`] = `
"function log() {
  return 'this should be in final bundle'
}

export { log as effect };
"
`;

exports[`node 20.12.1 - resolve-file-url > should resolve import with file:// annotation 2`] = `
"'use strict';

function log() {
  return 'this should be in final bundle'
}

exports.effect = log;
"
`;
